packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+///    intent to the [Router] widget, as well as reports new route information
+///    from the [Router] back to the engine by overriding the
+///    [routerReportsNewRouteInformation].
+abstract class RouteInformationProvider
+    extends ValueListenable<RouteInformation> {
+  /// A callback called when the [Router] widget detects any navigation event
+  /// due to state changes.
+  ///
+  /// The subclasses can override this method to update theirs values or trigger
+  /// other side effects. For example, the [PlatformRouteInformationProvider]
+  /// overrides this method to report the route information back to the engine.
+  ///
+  /// The [routeInformation] contains the route information after the
+  /// navigation event.
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {}
+}
+
+/// The route information provider that propagates the platform route information changes.
+///
+/// This provider also reports the new route information from the [Router] widget
+/// back to engine using message channel method, the
+/// [RouteNotificationMessages.maybeNotifyRouteChange].
+class PlatformRouteInformationProvider extends RouteInformationProvider
+    with WidgetsBindingObserver, ChangeNotifier {
+  /// Create a platform route information provider.
+  ///
+  /// Use the [initialRouteInformation] to set the default route information for this
+  /// provider.
+  PlatformRouteInformationProvider({RouteInformation initialRouteInformation})
+      : _value = initialRouteInformation,
+        _currentSerializedRouteInformationInEngine = _serializeRouteInformation(initialRouteInformation);
+
+  String _currentSerializedRouteInformationInEngine;
+
+  static String _serializeRouteInformation(RouteInformation routeInformation) {
+    String result;
+    try {
+      result = jsonEncode(<String, dynamic>{
+        'location': routeInformation.location,
+        'state': routeInformation.state,
+      });
+    } on JsonUnsupportedObjectError catch (error) {
+      FlutterError.reportError(
+        FlutterErrorDetails(
+          exception:
+          'The RouteInformation.state must be json serializable.\n' +
+            error.toString()),
+      );
+    }
+    return result;
+  }
+
+  static RouteInformation _deserializeRouteInformation(String serialized) {
+    final dynamic json = jsonDecode(serialized);
+    if (json is String)
+      return RouteInformation(location: json);
+    assert(json is Map<String, dynamic>);
+    final Map<String, dynamic> map = json as Map<String, dynamic>;
+    assert(map.containsKey('location') && map.containsKey('state'));
+    return RouteInformation(
+      location: json['location'] as String,
+      state: json['state'],
+    );
+  }
+
+  @override
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {
+    final String serializedRouteInformation = _serializeRouteInformation(routeInformation);
+    RouteNotificationMessages.maybeNotifyRouteChange(
```

Comment:
Shouldn't there be changes to RouteNotificationMessages as well? It currently assumes that the arguments are all simple route names and not what we pass in now.